### PR TITLE
Add `restrict(::Hist2D, ...)`

### PR DESCRIPTION
**Pull request by bojohnson5**
Hopefully this is a cleaner PR than my previous one. I ran the test I added and everything passed locally. I might try to add a slice range in the future instead of a single bin currently.

### CODE REVIEW

**Comment by Moelf** [comment]
I thought I understood this last time, but can you point me to an example of this function in another library (python `hist` or ROOT?).

It feels like this is equivalent to `profile` followed by a `restrict`?

**Comment by Moelf** [comment]
I guess here you end up with a histogram with same bin definitions but only one bin is populated, where in `profile` then `restrict` you have a histogram with only one bin.

**Comment by bojohnson5** [comment]
I was thinking of something along the lines of [ROOT](https://root.cern.ch/doc/master/classTH2.html#a974ece9e7d260f92df00a39dba14e5b0) where you project but only with a certain bin range.

**Comment by bojohnson5** [comment]
I'm not going for a profile here, per se, where I'm weighting over the other axis. I'm looking to take a restricted projection.

**Comment by Moelf** [comment]
say you start with a 2D histogram with 9 bins (and say the bincenters are just `1:3` in each axis)
```
julia> a = reshape(1:9, 3, 3)
3×3 reshape(::UnitRange{Int64}, 3, 3) with eltype Int64:
 1  4  7
 2  5  8
 3  6  9
```

you `project(a, :y, 2)` along y-axis would return something like:
```
[0, 15, 0] # as 1D hist
```
right?

**Comment by bojohnson5** [comment]
If you just do a `project(a, :y)` you get `[6, 15, 24]`. What I'm wanting to do is `project(a, :y, 2)` and get `[2, 5, 8]`. But now I'm realizing I can make the code much easier and cleaner than what I'm doing.

**Comment by bojohnson5** [comment]
I would ideally like to get a range in place of the 2 to allow a slice and not just a single bin

**Comment by Moelf** [comment]
> What I'm wanting to do is project(a, :y, 2) and get [2, 5, 8]

Oh, I see, then this is just a restrict for 2D histogram? (and if we say `restrict(::Hist2D) -> ::Hist2D`, we can still do a project afterwards to make it 1D

**Comment by Moelf** [comment]
currently there's no restrict for `Hist2D`, but I imagine the interface might be:
```
restrict(::Hist2D, xlow, xhi, ylow, yhigh)
```

**Comment by bojohnson5** [comment]
> > What I'm wanting to do is project(a, :y, 2) and get [2, 5, 8]
> 
> Oh, I see, then this is just a restrict for 2D histogram? (and if we say `restrict(::Hist2D) -> ::Hist2D`, we can still do a project afterwards to make it 1D

Not necessarily a restrict because let's say I did `project(a, :y, 1:2)` I'd like to to give a Hist1D that looked like `[3, 9, 15]`

**Comment by bojohnson5** [comment]
I'll look at implementing a `restrict` on Hist2D

**Comment by Moelf** [comment]
`project(a, :y, 1:2)` is still easily expressed in `restrict()` followed by `profile`. Basically I think in case of `Hist2D`, `restrict()` can always give you a `Hist2D` contains all the bins you want to "squash", and `profile()` always gives you the 1D histogram you want in the end.

**Comment by bojohnson5** [comment]
I think I've got things working how we want now. Had to make changes to the 1D chained restrict due to ambiguity with that and the 2D chained restrict.

**Comment by Moelf** [comment]
yeah the chains always seemed like a hack to me 😆

File: src/hist2d.jl [hunk 1]
```diff
@@ -313,3 +313,46 @@ function profile(h::Hist2D, axis::Symbol=:x)
...
+
+Returns a new histogram with a restricted x-axis.
+`restrict(h, 0, 3)` (or `h |> restrict(0, 3)`)
+will return a slice of `h` where the bin centers are in `[0, 3]` (inclusive).
+"""
+function restrict(h::Hist2D, xlow=-Inf, xhigh=Inf, ylow=-Inf, yhigh=Inf)
+    xsel = xlow .<= bincenters(h)[1] .<= xhigh
+    ysel = ylow .<= bincenters(h)[2] .<= yhigh
+    @assert sum(xsel) > 0 "No bin centers contained in [$(xlow), $(xhigh)]"
+    @assert sum(ysel) > 0 "No bin centers contained in [$(ylow), $(yhigh)]"
```
Comment (Moelf) [inline src/hist2d.jl:329]:
```suggestion
    @assert count(xsel) > 0 "No bin centers contained in [$(xlow), $(xhigh)]"
    @assert count(ysel) > 0 "No bin centers contained in [$(ylow), $(yhigh)]"
```

**Comment by Moelf** [comment]
looks good to me, I can merge if you're happy with this

**Review by Moelf — APPROVED** [review]

**Comment by bojohnson5** [comment]
Looks good to me!

**Pull request merged**